src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,396 @@
+        {
+            if (_isCompleted)
+            {
+                return;
+            }
+
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+
+            var segment = _readHead;
+            while (segment != null)
+            {
+                segment.ResetMemory();
+                segment = segment.NextSegment;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnWriterCompleted(Action<Exception, object> callback, object state)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            var tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out var readResult))
+            {
+                return readResult;
+            }
+
+            var reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeReader)state).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                var isCanceled = false;
+                try
+                {
+                    AllocateCommitHead();
+#if NETCOREAPP2_2
+                    var length = await _readingStream.ReadAsync(_commitHead.AvailableMemory, tokenSource.Token);
+#elif NETSTANDARD2_0
+                    if (!MemoryMarshal.TryGetArray<byte>(_commitHead.AvailableMemory, out var arraySegment))
+                    {
+                        throw new InvalidCastException("Could not get byte[] from Memory.");
+                    }
+
+                    var length = await _readingStream.ReadAsync(arraySegment.Array, 0, arraySegment.Count, tokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                    _commitHead.End += length;
+                    _consumedLength += length;
+                }
+                catch (OperationCanceledException)
+                {
+                    ClearOutCancellation();
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw;
+                    }
+
+                    isCanceled = true;
+                }
+
+                return new ReadResult(GetCurrentReadOnlySequence(), isCanceled, IsCompletedOrThrow());
+            }
+        }
+
+        private void ClearOutCancellation()
+        {
+            lock (lockObject)
+            {
+                _internalTokenSource = null;
+            }
+        }
+
+        private void ThrowIfCompleted()
```

Comment:
This methods won't inline due the `throw`. It would be better (?) to do the check for `_isCompleted` in the callee, so this method is just a "throwing" one. That keeps the non-completed path fast.

Or separate this method in one for the check only (can be inlined), that calls another one that throws if the check for _isCompleted_ is _true_.

Reply:
TIL. Very interesting.